### DISCUSSION TITLE
api 서버와 view단 uri 분리

### DISCUSSION THREAD

**Discussion by jangcheolhyeon** [BackendTheBone/MiraeCommerce#14] [General]
<h1>Trouble</h1>
api서버의 URL과 클라이언트의 URL을 어떻게 설정하는 것이 좋을까?

<h1>Trouble Shooting</h1>
정해진 규약이나 규칙이 없기 때문에 서비스기업들의 API를 가져왔습니다.

<h2>카카오 API</h2>
<a href="https://developers.kakao.com/docs/latest/ko/kakaologin/rest-api" >카카오 API 보기</a> <br>
<h3>카카오 로그인 API</h3>
curl -v -X POST "https://kauth.kakao.com/oauth/token" \ <br>
 -H "Content-Type: application/x-www-form-urlencoded" \ <br>
 -d "grant_type=authorization_code" \ <br>
 -d "client_id=${REST_API_KEY}" \ <br>
 --data-urlencode "redirect_uri=${REDIRECT_URI}" \ <br>
 -d "code=${AUTHORIZE_CODE}" <br>
<h3>사용자 목록 가져오는 API</h3>
https://kapi.kakao.com/v1/user/ids
<h3>동의 내역 확인하는 API</h3>
https://kapi.kakao.com/v2/user/scopes<hr>

<details>
<summary>참고 : 마이그레이션</summary>
<div markdown="1">

<strong>마이그레이션의 의미<strong>
IT 마이그레이션은 데이터나 소프트웨어를 한 시스템에서 다른 시스템으로 이동하는 것입니다. IT 마이그레이션은 프로젝트에 따라 데이터 마이그레이션, 애플리케이션 마이그레이션, 운영 체제 마이그레이션, 클라우드 마이그레이션 등 한 가지 이상의 이동이 진행될 수 있습니다.

IT 마이그레이션의 일반적인 몇 가지 예시는 다음과 같습니다.
<ul>
  <li>애플리케이션 또는 운영 체제(OS) 업그레이드</li>
  <li>데이터를 한 종류의 데이터베이스에서 다른 종류의 데이터베이스로 이동</li>
  <li>하나의 데이터 스토리지 시스템을 다른 데이터 스토리지 시스템으로 교체</li>
  <li>새로운 시스템이나 서비스를 마이그레이션을 통해 운영 및 유지보수 비용을 절감하고자 할때</li>
  <li>시스템의 성능이나 확장성이 제한적이거나 부족한 경우<li>
</ul>

위와 같이 api에 v1, v2가 들어가는 것을 볼 수 있는데 이는 v1보다 성능, 보안 등이 업데이트 된 경우 마이그레이션이라고 한다.

</div>
</details>

<h2>토스</h2>
<a href="https://docs.tosspayments.com/reference#%EA%B2%B0%EC%A0%9C" > 토스 API 보기</a> <br>
<h3>결제 승인 API</h3>
curl --request POST \ <br>
  --url https://api.tosspayments.com/v1/payments/confirm \ <br>
  --header 'Authorization: Basic dGVzdF9za196WExrS0V5cE5BcldtbzUwblgzbG1lYXhZRzVSOg==' \ <br>
  --header 'Content-Type: application/json' \ <br>
  --data '{"paymentKey":"5zJ4xY7m0kODnyRpQWGrN2xqGlNvLrKwv1M9ENjbeoPaZdL6","orderId":"a4CWyWY5m89PNh7xJwhk1","amount":15000}'
<br>

<h2>네이버</h2>
<a href="https://developers.naver.com/docs/common/openapiguide/apilist.md" >네이버 API 보기</a> <br>
<h3>네이버 로그인 인증 API</h3>
https://nid.naver.com/oauth2.0/authorize
<h3>카페 게시판에 게시글 등록 API</h3>
https://openapi.naver.com/v1/cafe/{clubid}/menu/{menuid}/articles<br>

<h1>OPINION</h1>
서비스 기업의 api 주소를 보면 Ouath를 제외한 api들은 url에 api가 들어가있는 것을 볼 수 있습니다.
규칙이나 규율이 있는 것은 아니지만 prefix를 통해 API 요청을 고유하게 식별함으로 사용하는 것이 좋다고 생각합니다.

**Comment by hamsteak1488** [comment]
정리해주신 자료를 포함해 유명한 서비스들의 API를 조사해 본 결과 많은 경우 uri를 'api.(도메인주소)/(버전)'과 같은 형식으로 사용하는 것 같네요. 저도 우리 프로젝트에 사용될 수 있는 uri를 제안해봅니다.

1. 'api.commerce.mirae.com/v1/'
2. 'api.commerce.mirae.com/'
3. 'commerce.mirae.com/api/v1/'
4. 'commerce.mirae.com/api/'
5. 'commerce.mirae.com/v1/'
